File: alejo/interaction/gesture_arpeggiator/static/js/audio-engine.js
/**
 * ALEJO Gesture Arpeggiator - Audio Engine
 * 
 * Handles client-side audio synthesis using Tone.js,
 * including arpeggiator and drum machine functionality.
 */

class AudioEngine {
    /**
     * Initialize the audio engine
     * @param {object} app - Reference to the main application
     */
    constructor(app) {
        this.app = app;
        
        // Audio state
        this.isInitialized = false;
        this.isPlaying = false;
        this.isMuted = false;
        
        // Audio components
        this.synth = null;
        this.drumSampler = null;
        this.arpeggioPattern = null;
        this.drumPattern = null;
        this.masterVolume = null;
        
        // Audio parameters
        this.bpm = 120;
        this.rootNote = 60; // Middle C
        this.scale = 'major';
        this.arpeggioPatternType = 'up';
        this.octaveRange = 2;
        this.currentDrumPatternId = 0;
        
        // Volume levels
        this.masterVolumeLevel = 0.8;
        this.arpeggioVolumeLevel = 0.8;
        this.drumVolumeLevel = 0.8;
        
        // Audio analysis
        this.analyzer = null;
        this.waveform = null;
        this.spectrum = null;
        
        // Musical data
        this.scales = {
            'major': [0, 2, 4, 5, 7, 9, 11],
            'minor': [0, 2, 3, 5, 7, 8, 10],
            'pentatonic': [0, 2, 4, 7, 9],
            'blues': [0, 3, 5, 6, 7, 10],
            'chromatic': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
        };
        
        // Drum patterns (kick, snare, hihat, tom)
        this.drumPatterns = [
            // Basic 4/4
            [
                [1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0],
                [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0],
                [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
                [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0]
            ],
            // Hip-hop
            [
                [1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0],
                [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1],
                [1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0],
                [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
            ],
            // Electronic
            [
                [1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0],
                [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0],
                [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
                [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0]
            ]
        ];
        
        // Bind methods
        this.onArpeggioStep = this.onArpeggioStep.bind(this);
        this.onDrumStep = this.onDrumStep.bind(this);
    }
    
    /**
     * Initialize the audio engine
     * @returns {Promise} Resolves when initialized
     */
    async initialize() {
        if (this.isInitialized) return;
        
        try {
            console.log('Initializing audio engine...');
            
            // Check if Tone.js is available
            if (!window.Tone) {
                throw new Error('Tone.js not available');
            }
            
            // Start audio context
            await Tone.start();
            
            // Set up master volume
            this.masterVolume = new Tone.Volume(Tone.gainToDb(this.masterVolumeLevel)).toDestination();
            
            // Set up synth
            this.synth = new Tone.PolySynth(Tone.Synth, {
                oscillator: {
                    type: 'triangle'
                },
                envelope: {
                    attack: 0.02,
                    decay: 0.1,
                    sustain: 0.3,
                    release: 1
                }
            }).connect(this.masterVolume);
            
            // Set up drum sampler
            this.drumSampler = new Tone.Sampler({
                urls: {
                    C1: 'kick.wav',
                    D1: 'snare.wav',
                    E1: 'hihat.wav',
                    F1: 'tom.wav'
                },
                baseUrl: '/static/audio/drums/',
                onload: () => {
                    console.log('Drum samples loaded');
                }
            }).connect(this.masterVolume);
            
            // Set up analyzer
            this.analyzer = new Tone.Analyser('waveform', 1024);
            this.masterVolume.connect(this.analyzer);
            
            // Set up spectrum analyzer
            this.spectrumAnalyzer = new Tone.Analyser('fft', 256);
            this.masterVolume.connect(this.spectrumAnalyzer);
            
            // Set BPM
            Tone.Transport.bpm.value = this.bpm;
            
            // Create arpeggio pattern
            this.createArpeggioPattern();
            
            // Create drum pattern
            this.createDrumPattern();
            
            this.isInitialized = true;
            console.log('Audio engine initialized');
            
            return true;
        } catch (error) {
            console.error('Failed to initialize audio engine:', error);
            this.isInitialized = false;
            throw error;
        }
    }
    
    /**
     * Start the audio engine
     * @returns {Promise} Resolves when started
     */
    async start() {
        if (this.isPlaying) return;
        
        try {
            // Initialize if not already
            if (!this.isInitialized) {
                await this.initialize();
            }
            
            console.log('Starting audio playback...');
            
            // Start transport
            Tone.Transport.start();
            
            this.isPlaying = true;
            console.log('Audio playback started');
            
            return true;
        } catch (error) {
            console.error('Failed to start audio playback:', error);
            this.isPlaying = false;
            throw error;
        }
    }
    
    /**
     * Stop the audio engine
     */
    stop() {
        if (!this.isPlaying) return;
        
        console.log('Stopping audio playback...');
        
        // Stop transport
        Tone.Transport.stop();
        
        this.isPlaying = false;
        console.log('Audio playback stopped');
    }
    
    /**
     * Toggle mute state
     * @returns {boolean} New mute state
     */
    toggleMute() {
        this.isMuted = !this.isMuted;
        
        if (this.masterVolume) {
            this.masterVolume.mute = this.isMuted;
        }
        
        console.log(`Audio ${this.isMuted ? 'muted' : 'unmuted'}`);
        return this.isMuted;
    }
    
    /**
     * Create arpeggio pattern
     */
    createArpeggioPattern() {
        // Clear existing pattern
        if (this.arpeggioPattern) {
            this.arpeggioPattern.dispose();
        }
        
        // Generate notes for the pattern
        const notes = this.generateArpeggioNotes();
        
        // Create new pattern
        this.arpeggioPattern = new Tone.Sequence(
            this.onArpeggioStep,
            notes,
            '16n'
        );
        
        // Start pattern
        this.arpeggioPattern.start(0);
    }
    
    /**
     * Generate notes for the arpeggio pattern
     * @returns {Array} Array of note frequencies
     */
    generateArpeggioNotes() {
        const scaleIntervals = this.scales[this.scale] || this.scales.major;
        const notes = [];
        
        // Generate notes based on pattern type
        switch (this.arpeggioPatternType) {
            case 'up':
                for (let octave = 0; octave < this.octaveRange; octave++) {
                    for (const interval of scaleIntervals) {
                        const note = this.rootNote + interval + (octave * 12);
                        notes.push(this.midiToFrequency(note));
                    }
                }
                break;
                
            case 'down':
                for (let octave = this.octaveRange - 1; octave >= 0; octave--) {
                    for (let i = scaleIntervals.length - 1; i >= 0; i--) {
                        const note = this.rootNote + scaleIntervals[i] + (octave * 12);
                        notes.push(this.midiToFrequency(note));
                    }
                }
                break;
                
            case 'upDown':
                // Up
                for (let octave = 0; octave < this.octaveRange; octave++) {
                    for (const interval of scaleIntervals) {
                        const note = this.rootNote + interval + (octave * 12);
                        notes.push(this.midiToFrequency(note));
                    }
                }
                // Down (excluding duplicates at the top)
                for (let octave = this.octaveRange - 1; octave >= 0; octave--) {
                    for (let i = scaleIntervals.length - 1; i >= 0; i--) {
                        if (octave === this.octaveRange - 1 && i === scaleIntervals.length - 1) {
                            continue; // Skip the first note in the descent (already played)
                        }
                        const note = this.rootNote + scaleIntervals[i] + (octave * 12);
                        notes.push(this.midiToFrequency(note));
                    }
                }
                break;
                
            case 'random':
                const totalNotes = scaleIntervals.length * this.octaveRange;
                for (let i = 0; i < totalNotes; i++) {
                    const octave = Math.floor(Math.random() * this.octaveRange);
                    const intervalIndex = Math.floor(Math.random() * scaleIntervals.length);
                    const note = this.rootNote + scaleIntervals[intervalIndex] + (octave * 12);
                    notes.push(this.midiToFrequency(note));
                }
                break;
                
            default:
                // Default to up
                for (let octave = 0; octave < this.octaveRange; octave++) {
                    for (const interval of scaleIntervals) {
                        const note = this.rootNote + interval + (octave * 12);
                        notes.push(this.midiToFrequency(note));
                    }
                }
        }
        
        return notes;
    }
    
    /**
     * Create drum pattern
     */
    createDrumPattern() {
        // Clear existing pattern
        if (this.drumPattern) {
            this.drumPattern.dispose();
        }
        
        // Create new pattern
        this.drumPattern = new Tone.Sequence(
            this.onDrumStep,
            Array.from({ length: 16 }, (_, i) => i),
            '16n'
        );
        
        // Start pattern
        this.drumPattern.start(0);
    }
    
    /**
     * Handle arpeggio step
     * @param {number} time - Time in seconds
     * @param {number} note - Note frequency
     */
    onArpeggioStep(time, note) {
        if (!this.isPlaying || this.isMuted) return;
        
        // Play note
        this.synth.triggerAttackRelease(note, '16n', time, this.arpeggioVolumeLevel);
        
        // Trigger event for visualization
        if (this.app && this.app.visualizer) {
            Tone.Draw.schedule(() => {
                this.app.visualizer.triggerNote(note);
            }, time);
        }
    }
    
    /**
     * Handle drum step
     * @param {number} time - Time in seconds
     * @param {number} step - Step index
     */
    onDrumStep(time, step) {
        if (!this.isPlaying || this.isMuted) return;
        
        const pattern = this.drumPatterns[this.currentDrumPatternId] || this.drumPatterns[0];
        
        // Play drum sounds
        if (pattern[0][step]) {
            this.drumSampler.triggerAttackRelease('C1', '16n', time, this.drumVolumeLevel);
        }
        
        if (pattern[1][step]) {
            this.drumSampler.triggerAttackRelease('D1', '16n', time, this.drumVolumeLevel);
        }
        
        if (pattern[2][step]) {
            this.drumSampler.triggerAttackRelease('E1', '16n', time, this.drumVolumeLevel * 0.7);
        }
        
        if (pattern[3][step]) {
            this.drumSampler.triggerAttackRelease('F1', '16n', time, this.drumVolumeLevel);
        }
        
        // Trigger event for visualization
        if (this.app && this.app.visualizer) {
            Tone.Draw.schedule(() => {
                this.app.visualizer.triggerDrum(step);
            }, time);
        }
    }
    
    /**
     * Convert MIDI note to frequency
     * @param {number} midi - MIDI note number
     * @returns {number} Frequency in Hz
     */
    midiToFrequency(midi) {
        return 440 * Math.pow(2, (midi - 69) / 12);
    }
    
    /**
     * Set master volume
     * @param {number} volume - Volume level (0-1)
     */
    setMasterVolume(volume) {
        this.masterVolumeLevel = volume;
        
        if (this.masterVolume) {
            this.masterVolume.volume.value = Tone.gainToDb(volume);
        }
    }
    
    /**
     * Set arpeggio volume
     * @param {number} volume - Volume level (0-1)
     */
    setArpeggioVolume(volume) {
        this.arpeggioVolumeLevel = volume;
    }
    
    /**
     * Set drum volume
     * @param {number} volume - Volume level (0-1)
     */
    setDrumVolume(volume) {
        this.drumVolumeLevel = volume;
    }
    
    /**
     * Set BPM
     * @param {number} bpm - Beats per minute
     */
    setBpm(bpm) {
        this.bpm = bpm;
        
        if (Tone.Transport) {
            Tone.Transport.bpm.value = bpm;
        }
    }
    
    /**
     * Set root note
     * @param {number} midiNote - MIDI note number
     */
    setRootNote(midiNote) {
        this.rootNote = midiNote;
        this.createArpeggioPattern();
    }
    
    /**
     * Set scale
     * @param {string} scale - Scale name
     */
    setScale(scale) {
        if (this.scales[scale]) {
            this.scale = scale;
            this.createArpeggioPattern();
        }
    }
    
    /**
     * Set arpeggio pattern
     * @param {string} pattern - Pattern type
     */
    setArpeggioPattern(pattern) {
        this.arpeggioPatternType = pattern;
        this.createArpeggioPattern();
    }
    
    /**
     * Set drum pattern
     * @param {number} patternId - Pattern ID
     */
    setDrumPattern(patternId) {
        if (patternId >= 0 && patternId < this.drumPatterns.length) {
            this.currentDrumPatternId = patternId;
        }
    }
    
    /**
     * Set octave range
     * @param {number} range - Octave range
     */
    setOctaveRange(range) {
        this.octaveRange = range;
        this.createArpeggioPattern();
    }
    
    /**
     * Update audio engine and return current state
     * @param {number} elapsed - Elapsed time since last update
     * @returns {object} Current audio state
     */
    update(elapsed) {
        // Get waveform data
        const waveform = this.analyzer ? this.analyzer.getValue() : new Float32Array(1024);
        
        // Get spectrum data
        const spectrum = this.spectrumAnalyzer ? this.spectrumAnalyzer.getValue() : new Float32Array(256);
        
        // Calculate audio levels
        let rms = 0;
        for (let i = 0; i < waveform.length; i++) {
            rms += waveform[i] * waveform[i];
        }
        rms = Math.sqrt(rms / waveform.length);
        
        // Calculate peak frequency
        let peakIndex = 0;
        let peakValue = 0;
        for (let i = 0; i < spectrum.length; i++) {
            if (spectrum[i] > peakValue) {
                peakValue = spectrum[i];
                peakIndex = i;
            }
        }
        
        const peakFrequency = peakIndex * (Tone.context.sampleRate / 2) / spectrum.length;
        
        // Return current state
        return {
            isPlaying: this.isPlaying,
            isMuted: this.isMuted,
            bpm: this.bpm,
            rootNote: this.rootNote,
            scale: this.scale,
            pattern: this.arpeggioPatternType,
            drumPattern: this.currentDrumPatternId,
            masterVolume: this.masterVolumeLevel,
            arpeggioVolume: this.arpeggioVolumeLevel,
            drumVolume: this.drumVolumeLevel,
            waveform: Array.from(waveform),
            spectrum: Array.from(spectrum),
            rms: rms,
            peakFrequency: peakFrequency,
            currentBeat: (Tone.Transport.position.split(':')[1] || 0)
        };
    }
}

File: alejo/interaction/gesture_arpeggiator/audio_engine.py
"""
Audio Engine for Gesture Arpeggiator

This module implements the audio generation engine for ALEJO's gesture-based
arpeggiator system, providing real-time synthesis of arpeggios and drum patterns
based on hand gesture controls.
"""

import logging
import asyncio
import json
import math
import time
from typing import Dict, List, Optional, Any, Union, Tuple
from enum import Enum
import numpy as np

logger = logging.getLogger(__name__)

class Scale(Enum):
    """Musical scales for the arpeggiator"""
    MAJOR = "major"
    MINOR = "minor"
    PENTATONIC = "pentatonic"
    BLUES = "blues"
    CHROMATIC = "chromatic"
    DORIAN = "dorian"
    PHRYGIAN = "phrygian"
    LYDIAN = "lydian"
    MIXOLYDIAN = "mixolydian"
    LOCRIAN = "locrian"

class ArpeggioPattern(Enum):
    """Arpeggio patterns"""
    UP = "up"
    DOWN = "down"
    UP_DOWN = "updown"
    DOWN_UP = "downup"
    RANDOM = "random"
    CONVERGE = "converge"
    DIVERGE = "diverge"
    PINKY_TO_THUMB = "pinky_to_thumb"

class AudioEngine:
    """
    Audio generation engine for the Gesture Arpeggiator
    
    This class handles the generation of musical notes and rhythms based on
    hand gesture controls, providing real-time synthesis for the arpeggiator
    and drum machine.
    """
    
    def __init__(self):
        """Initialize the audio engine"""
        self._active = False
        self._bpm = 120
        self._current_beat = 0
        self._beat_interval = 60.0 / self._bpm / 4  # 16th note interval
        self._last_beat_time = 0
        
        # Musical parameters
        self._root_note = 60  # Middle C in MIDI
        self._scale_type = Scale.MAJOR
        self._octave_range = 2
        self._arpeggio_pattern = ArpeggioPattern.UP
        self._arpeggio_notes = []
        self._current_arpeggio_step = 0
        
        # Drum parameters
        self._drum_pattern_id = 0
        self._drum_patterns = self._initialize_drum_patterns()
        self._current_drum_hits = {"kick": False, "snare": False, "hihat": False, "tom": False}
        
        # Volume controls
        self._master_volume = 0.8
        self._arpeggio_volume = 0.8
        self._drum_volume = 0.8
        
        # Audio generation state
        self._audio_buffer = np.zeros(1024, dtype=np.float32)
        self._sample_rate = 44100
        
        # Scale definitions (semitone intervals from root)
        self._scales = {
            Scale.MAJOR: [0, 2, 4, 5, 7, 9, 11],
            Scale.MINOR: [0, 2, 3, 5, 7, 8, 10],
            Scale.PENTATONIC: [0, 2, 4, 7, 9],
            Scale.BLUES: [0, 3, 5, 6, 7, 10],
            Scale.CHROMATIC: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
            Scale.DORIAN: [0, 2, 3, 5, 7, 9, 10],
            Scale.PHRYGIAN: [0, 1, 3, 5, 7, 8, 10],
            Scale.LYDIAN: [0, 2, 4, 6, 7, 9, 11],
            Scale.MIXOLYDIAN: [0, 2, 4, 5, 7, 9, 10],
            Scale.LOCRIAN: [0, 1, 3, 5, 6, 8, 10]
        }
        
        # Initialize arpeggio notes
        self._update_arpeggio_notes()
        
    def _initialize_drum_patterns(self) -> List[Dict]:
        """Initialize predefined drum patterns"""
        return [
            {
                "name": "Four on the floor",
                "kick": [1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0],
                "snare": [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0],
                "hihat": [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
                "tom": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
            },
            {
                "name": "Hip Hop",
                "kick": [1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0],
                "snare": [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0],
                "hihat": [1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0],
                "tom": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
            },
            {
                "name": "Breakbeat",
                "kick": [1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0],
                "snare": [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0],
                "hihat": [1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1],
                "tom": [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
            },
            {
                "name": "Techno",
                "kick": [1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0],
                "snare": [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
                "hihat": [0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1],
                "tom": [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
            },
            {
                "name": "Jazz",
                "kick": [1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0],
                "snare": [0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0],
                "hihat": [1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
                "tom": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
            }
        ]
        
    def start(self):
        """Start the audio engine"""
        if not self._active:
            self._active = True
            self._last_beat_time = time.time()
            logger.info("Audio engine started")
            
    def stop(self):
        """Stop the audio engine"""
        if self._active:
            self._active = False
            logger.info("Audio engine stopped")
            
    def is_active(self) -> bool:
        """Check if the audio engine is active"""
        return self._active
        
    def update(self) -> Dict:
        """
        Update the audio engine state
        
        Returns:
            Dictionary with current audio state
        """
        if not self._active:
            return self._get_audio_state()
            
        current_time = time.time()
        elapsed = current_time - self._last_beat_time
        
        # Check if it's time for the next beat
        if elapsed >= self._beat_interval:
            # Calculate how many beats have passed (in case of lag)
            beats_passed = int(elapsed / self._beat_interval)
            self._current_beat = (self._current_beat + beats_passed) % 16
            self._last_beat_time = current_time
            
            # Update arpeggio step
            self._update_arpeggio_step()
            
            # Update drum hits
            self._update_drum_hits()
            
            # Generate audio for this beat
            self._generate_audio()
            
        return self._get_audio_state()
        
    def _update_arpeggio_step(self):
        """Update the current arpeggio step based on pattern"""
        if not self._arpeggio_notes:
            return
            
        pattern = self._arpeggio_pattern
        num_notes = len(self._arpeggio_notes)
        
        if pattern == ArpeggioPattern.UP:
            self._current_arpeggio_step = (self._current_arpeggio_step + 1) % num_notes
        elif pattern == ArpeggioPattern.DOWN:
            self._current_arpeggio_step = (self._current_arpeggio_step - 1) % num_notes
        elif pattern == ArpeggioPattern.UP_DOWN:
            # Up and down pattern
            cycle_length = 2 * num_notes - 2
            step = (self._current_beat % cycle_length)
            if step < num_notes:
                self._current_arpeggio_step = step
            else:
                self._current_arpeggio_step = 2 * num_notes - 2 - step
        elif pattern == ArpeggioPattern.RANDOM:
            import random
            self._current_arpeggio_step = random.randint(0, num_notes - 1)
        elif pattern == ArpeggioPattern.CONVERGE:
            # Notes converge from outside to center
            half = num_notes // 2
            if self._current_beat % 2 == 0:
                self._current_arpeggio_step = half - (self._current_beat // 2) % half
            else:
                self._current_arpeggio_step = half + ((self._current_beat + 1) // 2) % half
        elif pattern == ArpeggioPattern.DIVERGE:
            # Notes diverge from center to outside
            half = num_notes // 2
            if self._current_beat % 2 == 0:
                self._current_arpeggio_step = half + (self._current_beat // 2) % half
            else:
                self._current_arpeggio_step = half - ((self._current_beat + 1) // 2) % half
                
    def _update_drum_hits(self):
        """Update drum hits based on current beat and pattern"""
        if self._drum_pattern_id >= len(self._drum_patterns):
            self._drum_pattern_id = 0
            
        pattern = self._drum_patterns[self._drum_pattern_id]
        beat_idx = self._current_beat % 16
        
        self._current_drum_hits = {
            "kick": pattern["kick"][beat_idx] == 1,
            "snare": pattern["snare"][beat_idx] == 1,
            "hihat": pattern["hihat"][beat_idx] == 1,
            "tom": pattern["tom"][beat_idx] == 1
        }
        
    def _generate_audio(self):
        """Generate audio for the current beat"""
        # In a real implementation, this would generate actual audio samples
        # For now, we'll just update the state
        pass
        
    def _get_audio_state(self) -> Dict:
        """
        Get the current audio state
        
        Returns:
            Dictionary with current audio parameters
        """
        current_note = None
        if self._arpeggio_notes and 0 <= self._current_arpeggio_step < len(self._arpeggio_notes):
            current_note = self._arpeggio_notes[self._current_arpeggio_step]
            
        return {
            "active": self._active,
            "bpm": self._bpm,
            "current_beat": self._current_beat,
            "current_note": current_note,
            "drum_hits": self._current_drum_hits,
            "master_volume": self._master_volume,
            "arpeggio_volume": self._arpeggio_volume,
            "drum_volume": self._drum_volume
        }
        
    def _update_arpeggio_notes(self):
        """Update the arpeggio notes based on current settings"""
        scale = self._scales[self._scale_type]
        notes = []
        
        # Generate notes across octave range
        for octave in range(self._octave_range):
            for interval in scale:
                note = self._root_note + interval + (octave * 12)
                notes.append(note)
                
        self._arpeggio_notes = notes
        self._current_arpeggio_step = 0
        
    def set_bpm(self, bpm: int):
        """
        Set the tempo in beats per minute
        
        Args:
            bpm: Tempo in beats per minute
        """
        if 40 <= bpm <= 300:
            self._bpm = bpm
            self._beat_interval = 60.0 / self._bpm / 4  # 16th note interval
            
    def set_root_note(self, midi_note: int):
        """
        Set the root note for the arpeggiator
        
        Args:
            midi_note: MIDI note number (0-127)
        """
        if 0 <= midi_note <= 127:
            self._root_note = midi_note
            self._update_arpeggio_notes()
            
    def set_scale(self, scale_type: Scale):
        """
        Set the scale type for the arpeggiator
        
        Args:
            scale_type: Scale type enum
        """
        if scale_type in Scale:
            self._scale_type = scale_type
            self._update_arpeggio_notes()
            
    def set_octave_range(self, octave_range: int):
        """
        Set the octave range for the arpeggiator
        
        Args:
            octave_range: Number of octaves (1-4)
        """
        if 1 <= octave_range <= 4:
            self._octave_range = octave_range
            self._update_arpeggio_notes()
            
    def set_arpeggio_pattern(self, pattern: ArpeggioPattern):
        """
        Set the arpeggio pattern
        
        Args:
            pattern: Arpeggio pattern enum
        """
        if pattern in ArpeggioPattern:
            self._arpeggio_pattern = pattern
            
    def set_drum_pattern(self, pattern_id: int):
        """
        Set the drum pattern
        
        Args:
            pattern_id: Index of the drum pattern
        """
        if 0 <= pattern_id < len(self._drum_patterns):
            self._drum_pattern_id = pattern_id
            
    def set_master_volume(self, volume: float):
        """
        Set the master volume
        
        Args:
            volume: Volume level (0.0-1.0)
        """
        self._master_volume = max(0.0, min(1.0, volume))
        
    def set_arpeggio_volume(self, volume: float):
        """
        Set the arpeggiator volume
        
        Args:
            volume: Volume level (0.0-1.0)
        """
        self._arpeggio_volume = max(0.0, min(1.0, volume))
        
    def set_drum_volume(self, volume: float):
        """
        Set the drum volume
        
        Args:
            volume: Volume level (0.0-1.0)
        """
        self._drum_volume = max(0.0, min(1.0, volume))
        
    def get_available_scales(self) -> Dict[str, List[int]]:
        """
        Get available musical scales
        
        Returns:
            Dictionary of scale names and intervals
        """
        return {scale.value: intervals for scale, intervals in self._scales.items()}
        
    def get_available_patterns(self) -> List[str]:
        """
        Get available arpeggio patterns
        
        Returns:
            List of pattern names
        """
        return [pattern.value for pattern in ArpeggioPattern]
        
    def get_available_drum_patterns(self) -> List[Dict]:
        """
        Get available drum patterns
        
        Returns:
            List of drum pattern dictionaries
        """
        return self._drum_patterns
        
    def note_to_frequency(self, midi_note: int) -> float:
        """
        Convert MIDI note number to frequency in Hz
        
        Args:
            midi_note: MIDI note number (0-127)
            
        Returns:
            Frequency in Hz
        """
        return 440.0 * (2.0 ** ((midi_note - 69) / 12.0))
        
    def note_to_name(self, midi_note: int) -> str:
        """
        Convert MIDI note number to note name
        
        Args:
            midi_note: MIDI note number (0-127)
            
        Returns:
            Note name (e.g., "C4")
        """
        notes = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]
        octave = (midi_note // 12) - 1
        note = notes[midi_note % 12]
        return f"{note}{octave}"
